README.md
# Oh Dear Python

[![Latest Version on PyPi][ico-version]][link-pypi]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-github-actions]][link-github-actions]
[![Buy us a tree][ico-treeware-gifting]][link-treeware-gifting]

An SDK to easily work with the [Oh Dear API](https://ohdear.app/docs/integrations/the-oh-dear-api)

## Install

Via Pip

```shell
pip install ohdear-sdk
```

## Usage

```python
from ohdear import OhDear

ohdear = OhDear(api_token="your-token")

sites = ohdear.sites.all()
```

| Available Methods                      | Description                                                                                         |
|:---------------------------------------|:----------------------------------------------------------------------------------------------------|
| `ohdear.me()`                          | Retrieve a `UserInfo` dict with details about the currently authenticated user.                     |
| `ohdear.authenticated()`               | Retrieve a `boolean` response indicating whether the current user is authenticated.                 |
| `ohdear.broken_links.show(123)`        | Retrieve a `BrokenLinksCollection` dict with details about broken links for a specific site.        |
| `ohdear.certificates.enable(123)`      | Retrieve a `CertificateHealth` dict containing details about the certificate and it's health.       |
| `ohdear.checks.enable(123)`            | Retrieve a `boolean` indicating whether a check was enabled successfully.                           |
| `ohdear.checks.disable(123)`           | Retrieve a `boolean` indicating whether a check was disabled successfully.                          |
| `ohdear.cron_checks.show(123)`         | Retrieve a `CronChecksCollection` dict with details about the cron checks for a specific site.      |
| `ohdear.domain_monitoring.domain(123)` | Retrieve a `Domain` dict with details about the domain for a specific site.                         |
| `ohdear.mixed_contents.show(123)`      | Retrieve a `MixedContentsCollection` dict with details about the mixed content for a specific site. |
| `ohdear.sites.all()`                   | Retrieve a `SitesCollection` dict with details about all sites.                                     |
| `ohdear.sites.show(123)`               | Retrieve a `Site` dict with details about a specific site.                                          |

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

```shell
hatch shell

hatch run test
```

## Security

If you discover any security related issues, please email security@voke.dev instead of using the issue tracker.

## Credits

- [Owen Voke][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

## Treeware

You're free to use this package, but if it makes it to your production environment you are required to buy the world a tree.

It’s now common knowledge that one of the best tools to tackle the climate crisis and keep our temperatures from rising above 1.5C is to plant trees. If you support this package and contribute to the Treeware forest you’ll be creating employment for local families and restoring wildlife habitats.

You can buy trees [here][link-treeware-gifting].

Read more about Treeware at [treeware.earth][link-treeware].

[ico-version]: https://img.shields.io/pypi/v/ohdear-sdk.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-github-actions]: https://img.shields.io/github/actions/workflow/status/owenvoke/ohdear-python-sdk/tests.yml?branch=main&style=flat-square
[ico-treeware-gifting]: https://img.shields.io/badge/Treeware-%F0%9F%8C%B3-lightgreen?style=flat-square

[link-pypi]: https://pypi.org/project/ohdear-sdk
[link-github-actions]: https://github.com/owenvoke/ohdear-python-sdk/actions
[link-treeware]: https://treeware.earth
[link-treeware-gifting]: https://ecologi.com/owenvoke?gift-trees
[link-author]: https://github.com/owenvoke
[link-contributors]: https://github.com/owenvoke/ohdear-python-sdk/contributors


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com), and this project adheres to [Semantic Versioning](https://semver.org).

## Unreleased

## [v0.3.9 - 2023-04-18](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.8...v0.3.9)

### Added
- Add support for domain monitoring ([#4](https://github.com/owenvoke/ohdear-python-sdk/pull/4))

## [v0.3.8 - 2022-11-18](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.7...v0.3.8)

### Added
- Enforce types with `py.typed` file ([#3](https://github.com/owenvoke/ohdear-python-sdk/pull/3))

## [v0.3.7 - 2022-11-16](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.6...v0.3.7)

### Added
- Add Hatch / PyProject configuration ([#2](https://github.com/owenvoke/ohdear-python-sdk/pull/2))

## [v0.3.6 - 2022-11-06](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.5...v0.3.6)

### Added
- Export exceptions in package

## [v0.3.5 - 2022-11-06](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.4...v0.3.5)

### Added
- Add missing Check `summary` property

## [v0.3.4 - 2022-11-06](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.3...v0.3.4)

### Added
- Export types in package ([#1](https://github.com/owenvoke/ohdear-python-sdk/pull/1))

## [v0.3.3 - 2022-10-22](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.2...v0.3.3)

### Added
- Add support for mixed contents

## [v0.3.2 - 2022-10-21](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.1...v0.3.2)

### Added
- Add support for certificates

## [v0.3.1 - 2022-10-21](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.3.0...v0.3.1)

### Changed
- Added long description for PyPi

## [v0.3.0 - 2022-10-21](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.2.0...v0.3.0)

### Added
- Add support for checks
- Add support for cron checks
- Add support for broken links

### Changed
- **BREAKING**: Moved types from `ohdear.models` to `ohdear.types`

## [v0.2.0 - 2022-10-21](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.1.1...v0.2.0)

### Added
- Add stricter typing and tests

### Changed
- **BREAKING**: Added SitesCollection `TypedDict` (sites are now under a `data` key)

## [v0.1.1 - 2022-10-20](https://github.com/owenvoke/ohdear-python-sdk/compare/v0.1.0...v0.1.1)

### Added
- Add support for `/me`

## v0.1.0 - 2022-10-20

### Added
- Initial release


tests/test_sites.py
import pytest
import requests_mock

from ohdear import NotFoundException, OhDear

ohdear = OhDear("abcdefghijklmnopqrstuvwxyz123457890")


def test_sites_all():
    with requests_mock.Mocker() as m:
        m.get(
            url="https://ohdear.app/api/sites",
            json={
                "data": [
                    {
                        "id": 1,
                        "url": "https://yoursite.tld",
                        "sort_url": "yoursite.tld",
                        "label": "your-site",
                        "team_id": 1,
                        "group_name": None,
                        "latest_run_date": "2019-09-16 07:29:02",
                        "summarized_check_result": "succeeded",
                        "created_at": "20171106 07:40:49",
                        "updated_at": "20171106 07:40:49",
                        "checks": [
                            {
                                "id": 100,
                                "type": "uptime",
                                "label": "Uptime",
                                "enabled": True,
                                "latest_run_ended_at": "2019-09-16 07:29:02",
                                "latest_run_result": "succeeded",
                                "summary": "Up",
                            }
                        ],
                    }
                ]
            },
        )
        all_sites = ohdear.sites.all()

        assert type(all_sites) == dict
        assert type(all_sites.get("data")) == list
        assert type(all_sites.get("data")[0]) == dict
        assert type(all_sites.get("data")[0].get("id")) == int


def test_sites_single():
    with requests_mock.Mocker() as m:
        m.get(
            url="https://ohdear.app/api/sites/12345",
            json={
                "id": 1,
                "url": "https://yoursite.tld",
                "sort_url": "yoursite.tld",
                "label": "your-site",
                "team_id": 1,
                "group_name": None,
                "latest_run_date": "2019-09-16 07:29:02",
                "summarized_check_result": "succeeded",
                "created_at": "20171106 07:40:49",
                "updated_at": "20171106 07:40:49",
                "checks": [
                    {
                        "id": 100,
                        "type": "uptime",
                        "label": "Uptime",
                        "enabled": True,
                        "latest_run_ended_at": "2019-09-16 07:29:02",
                        "latest_run_result": "succeeded",
                        "summary": "Up",
                    }
                ],
            },
        )
        site = ohdear.sites.show(12345)

        assert type(site) == dict
        assert type(site.get("id")) == int


def test_sites_single_for_a_non_existent_site():
    with requests_mock.Mocker() as m:
        m.get(
            url="https://ohdear.app/api/sites/1",
            status_code=404,
            json={"message": "No query results for model [no-model]."},
        )
        with pytest.raises(NotFoundException):
            assert ohdear.sites.show(1)


tests/test_checks.py
import requests_mock

from ohdear import OhDear

ohdear = OhDear("abcdefghijklmnopqrstuvwxyz123457890")


def test_enables_check():
    with requests_mock.Mocker() as m:
        m.post(
            url="https://ohdear.app/api/checks/123/enable",
            json={"id": 405, "type": "uptime", "enabled": True},
        )
        status = ohdear.checks.enable(123)

        assert type(status) == bool
        assert status is True


def test_disables_check():
    with requests_mock.Mocker() as m:
        m.post(
            url="https://ohdear.app/api/checks/123/disable",
            json={"id": 405, "type": "uptime", "enabled": False},
        )
        status = ohdear.checks.disable(123)

        assert type(status) == bool
        assert status is False


tests/test_ohdear.py
import requests_mock

from ohdear import OhDear

ohdear = OhDear("abcdefghijklmnopqrstuvwxyz123457890")


def test_me():
    with requests_mock.Mocker() as m:
        m.get(
            url="https://ohdear.app/api/me",
            json={
                "id": 1,
                "name": "Owen Voke",
                "email": "development@voke.dev",
                "photo_url": "https://example.org/profile-1.jpg",
                "teams": [{"id": 1, "name": "Dream Team"}],
            },
        )
        me = ohdear.me()

        assert type(me) == dict
        assert me.get("id") == 1
        assert me.get("name") == "Owen Voke"
        assert me.get("email") == "development@voke.dev"
        assert me.get("photo_url") == "https://example.org/profile-1.jpg"
        assert me.get("teams")[0].get("id") == 1
        assert me.get("teams")[0].get("name") == "Dream Team"


def test_authenticated():
    with requests_mock.Mocker() as m:
        m.get(
            url="https://ohdear.app/api/me",
            json={
                "id": 1,
                "name": "Owen Voke",
                "email": "development@voke.dev",
                "photo_url": "https://example.org/profile-1.jpg",
                "teams": [{"id": 1, "name": "Dream Team"}],
            },
        )
        me = ohdear.authenticated()

        assert type(me) == bool
        assert me is True


def test_not_authenticated():
    with requests_mock.Mocker() as m:
        m.get(
            url="https://ohdear.app/api/me",
            status_code=401,
            json={"error": "Unauthorized."},
        )
        me = ohdear.authenticated()

        assert type(me) == bool
        assert me is False


ohdear/types.py
from datetime import datetime
from typing import Optional, TypedDict


class Team(TypedDict):
    id: int
    name: str


class UserInfo(TypedDict):
    id: int
    name: str
    email: str
    photo_url: str
    teams: list[Team]


class Check(TypedDict):
    id: int
    type: str
    label: str
    enabled: bool
    latest_run_ended_at: datetime
    latest_run_result: str
    summary: str


class Site(TypedDict):
    id: int
    url: str
    sort_url: str
    friendly_name: Optional[str]
    label: str
    team_id: int
    latest_run_date: Optional[datetime]
    summarized_check_result: str
    uses_https: Optional[bool]
    checks: list[Check]
    broken_links_check_include_external_links: Optional[bool]
    broken_links_whitelisted_urls: Optional[list[str]]
    created_at: Optional[datetime]
    update_at: Optional[datetime]


class SitesCollection(TypedDict):
    data: list[Site]


class CronCheck(TypedDict):
    id: int
    uuid: str
    name: str
    type: str
    description: str
    frequency_in_minutes: str
    grace_time_in_minutes: str
    cron_expression: Optional[str]
    server_timezone: str
    ping_url: str
    created_at: datetime
    latest_result: str
    latest_ping_at: Optional[datetime]


class CronChecksCollection(TypedDict):
    data: list[CronCheck]


class BrokenLink(TypedDict):
    crawled_url: str
    status_code: int
    found_on_url: str


class BrokenLinksCollection(TypedDict):
    data: list[BrokenLink]


class CertificateDetails(TypedDict):
    issuer: str
    valid_from: datetime
    valid_until: datetime


class CertificateCheck(TypedDict):
    type: str
    label: str
    passed: bool


class CertificateHealth(TypedDict):
    certificate_details: CertificateDetails
    certificate_checks: list[CertificateCheck]
    certificate_chain_issuers: list[str]


class Domain(TypedDict):
    expires_at: datetime
    registered_at: datetime
    last_changed_at: datetime
    last_updated_in_rdap_db_at: datetime
    domain_statuses: dict[str, bool]
    rdap_domain_response: dict
    created_at: datetime


class MixedContent(TypedDict):
    element_name: str
    mixed_content_url: str
    found_on_url: str


class MixedContentsCollection(TypedDict):
    data: list[MixedContent]


ohdear/models.py
from datetime import datetime
from typing import TypedDict, Optional


class Team(TypedDict):
    id: int
    name: str


class UserInfo(TypedDict):
    id: int
    name: str
    email: str
    photo_url: str
    teams: list[Team]


class Check(TypedDict):
    id: int
    type: str
    label: str
    enabled: bool
    latest_run_ended_at: datetime
    latest_run_result: str


class Site(TypedDict):
    id: int
    url: str
    sort_url: str
    friendly_name: Optional[str]
    label: str
    team_id: int
    latest_run_date: Optional[datetime]
    summarized_check_result: str
    uses_https: Optional[bool]
    checks: list[Check]
    broken_links_check_include_external_links: Optional[bool]
    broken_links_whitelisted_urls: Optional[list[str]]
    created_at: Optional[datetime]
    update_at: Optional[datetime]


class SitesCollection(TypedDict):
    data: list[Site]


class CronCheck(TypedDict):
    id: int
    uuid: str
    name: str
    type: str
    description: str
    frequency_in_minutes: str
    grace_time_in_minutes: str
    cron_expression: Optional[str]
    server_timezone: str
    ping_url: str
    created_at: datetime
    latest_result: str
    latest_ping_at: Optional[datetime]


class CronChecksCollection(TypedDict):
    data: list[Site]


ohdear/ohdear.py
from typing import Optional, cast

import requests

from ohdear.types import (
    BrokenLinksCollection,
    CertificateHealth,
    CronChecksCollection,
    Domain,
    MixedContentsCollection,
    Site,
    SitesCollection,
    UserInfo,
)

TIMEOUT = 3
API_BASE_URI = "https://ohdear.app/api"


class OhDearException(RuntimeError):
    """Unknown error"""


class UnauthorizedException(OhDearException):
    """Unauthorized"""


class NotFoundException(OhDearException):
    """Not Found"""


class OhDear:
    def __init__(self, api_token: str, base_uri: str = API_BASE_URI) -> None:
        self.api_token: str = api_token
        self.base_uri: str = base_uri
        self.headers: dict = {"Authorization": "Bearer {0}".format(self.api_token)}

        self.broken_links: BrokenLinks = BrokenLinks(self)
        self.certificates: Certificates = Certificates(self)
        self.checks: Checks = Checks(self)
        self.cron_checks: CronChecks = CronChecks(self)
        self.domain_monitoring: DomainMonitoring = DomainMonitoring(self)
        self.mixed_contents: MixedContents = MixedContents(self)
        self.sites: Sites = Sites(self)

    def authenticated(self) -> bool:
        try:
            return self.me().get("id") is not None
        except UnauthorizedException:
            return False

    def me(self) -> UserInfo:
        return cast(UserInfo, self.get("/me"))

    def get(self, url: str):
        return self.__request("GET", url)

    def post(self, url: str, data: Optional[dict] = None):
        if data is None:
            data = {}
        return self.__request("POST", url, data)

    def __request(self, method: str, url: str, params: Optional[dict] = None):
        if params is None:
            params = {}

        if method == "GET":
            response = requests.get(
                self.base_uri + url, params=params, headers=self.headers
            )
        elif method == "POST":
            response = requests.post(
                self.base_uri + url, json=params, headers=self.headers
            )
        else:
            raise RuntimeError("Invalid request method provided")

        if response.status_code == 401:
            raise UnauthorizedException(response.json().get("error"))
        if response.status_code == 404:
            raise NotFoundException(response.json().get("error"))
        if response.status_code >= 400:
            raise OhDearException(response.json().get("error") or "Unknown error")
        return response.json()


class BrokenLinks:
    def __init__(self, client: OhDear):
        self.client = client

    def show(self, site_id: int) -> BrokenLinksCollection:
        return cast(BrokenLinksCollection, self.client.get(f"/broken-links/{site_id}"))


class Certificates:
    def __init__(self, client: OhDear):
        self.client = client

    def show(self, site_id: int) -> CertificateHealth:
        return cast(
            CertificateHealth, self.client.get(f"/certificate-health/{site_id}")
        )


class Checks:
    def __init__(self, client: OhDear):
        self.client = client

    def enable(self, check_id: int) -> bool:
        return self.client.post(f"/checks/{check_id}/enable").get("enabled")

    def disable(self, check_id: int) -> bool:
        return self.client.post(f"/checks/{check_id}/disable").get("enabled")


class CronChecks:
    def __init__(self, client: OhDear):
        self.client = client

    def show(self, site_id: int) -> CronChecksCollection:
        return cast(
            CronChecksCollection, self.client.get(f"/sites/{site_id}/cron-checks")
        )


class DomainMonitoring:
    def __init__(self, client: OhDear):
        self.client = client

    def domain(self, site_id: int) -> Domain:
        return cast(Domain, self.client.get(f"/sites/{site_id}/domain"))


class MixedContents:
    def __init__(self, client: OhDear):
        self.client = client

    def show(self, site_id: int) -> MixedContentsCollection:
        return cast(
            MixedContentsCollection, self.client.get(f"/mixed-content/{site_id}")
        )


class Sites:
    def __init__(self, client: OhDear):
        self.client = client

    def all(self) -> SitesCollection:
        return cast(SitesCollection, self.client.get("/sites"))

    def show(self, site_id: int) -> Site:
        return cast(Site, self.client.get(f"/sites/{site_id}"))
